### ISSUE TITLE
Produce standalone fonts when subsetting

### ISSUE THREAD

**Issue by wezm** [yeslogic/allsorts#27]
The subsetting feature is currently tailored for the needs of subsetting fonts for embedding in PDFs, since that was our primary use case when developing Allsorts. The issue is that we don't include a cmap table in the subset font, which makes it invalid for use outside PDF. When a subset font is embedded in a PDF the cmap info is contained in the PDF directly, so we don't need to include it in the font.

In order to support more general subsetting it would be convenient to have an entry point that takes a list of chars and produces a font with glyphs for just those chars. This would be an incremental improvement on what we have so far and would still have some limitations: with chars as input there wouldn't be a way to include ligature glyphs. Doing so would require subsetting `gpos`, and `gsub` tables as well, which is a problem for another day.

The subsetting code lives in [subset.rs](https://github.com/yeslogic/allsorts/blob/master/src/subset.rs). The new function signature could be along these lines:

```rust
/// Subset this font so that it only contains the glyphs for the supplied `chars`.
pub fn subset_chars(
    provider: &impl FontTableProvider,
    chars: &[char],
) -> Result<Vec<u8>, ReadWriteError>
```

The implementation would need to map `chars` to glyph ids using a technique [similar to this](https://github.com/yeslogic/allsorts-tools/blob/a7f6efb525657c7a0eaf968688c3422e3589ff27/src/subset.rs#L62). The subset font would need to include a new [cmap table](https://docs.microsoft.com/en-us/typography/opentype/spec/cmap) (probably using the [Unicode platformID](https://docs.microsoft.com/en-us/typography/opentype/spec/cmap#encoding-records-and-encodings)). There's a bunch of formats to chose from to encode the data. An initial implementation might just choose one of the simpler ones at the cost of size of the resulting font. A more sophisticated implementation could examine the data to determine the best option.

**Comment by ebraminio** [comment]
> The issue is that we don't include a cmap table in the subset font, which makes it invalid for use outside PDF.

Some PDF readers also won't work without a valid `cmap`, https://crbug.com/1071958 guess is needed for their text selection to work properly.

**Comment by wezm** [comment]
I've just released 0.9, which implements building of a proper `cmap` table for subset fonts.

**Comment by yisibl** [comment]
@wezm Can you upgrade the dependency version in allsorts-tools?

Looks like it can be solved: https://github.com/yeslogic/allsorts-tools/issues/16

**Comment by wezm** [comment]
Yes I'm working on that next. I have a draft PR open for it https://github.com/yeslogic/allsorts-tools/pull/18

**Comment by wezm** [comment]
Reopening as we strip the `OS/2` table which is required in OpenType fonts.

**Comment by yisibl** [comment]
@wezm I tried to submit a PR to fix it, PTAL. https://github.com/yeslogic/allsorts/pull/58

**Comment by yisibl** [comment]
Happy New Year! Any progress here?

**Comment by wezm** [comment]
No, sorry it's a pretty big piece of work that has not been scheduled yet.

**Comment by dnlmlr** [comment]
Hey! I am also trying to use subsetting for embedded fonts in PDF documents. Since I want to avoid getting too deep into the low level PDF structure, I am just using the `genpdf` -> `printpdf` -> `lopdf` stack. The plan was to embed the full subsetted font into the PDF files without touching the PDF internal mappings `/Differences`. 

I got it to work on all tested PDF readers and printers with the current implementation of `subset` even though the `OS/2` table is missing, but only if Unicode Encoding Records are used (mappings with `CharExistence::BasicMultilingualPlane`, `CharExistence::AstralPlane`). If `CharExistence::MacRoman` or `CharExistence::DivinePlane` is used, it doesn't work.

Would it be a sensible thing to allow forcing the default mode to be Unicode or are there any problems with this?

**Comment by wezm** [comment]
> Would it be a sensible thing to allow forcing the default mode to be Unicode or are there any problems with this?

I don't think that would make sense as a default as it would unnecessarily inflate the font. There is already an internal [CmapStrategy](https://github.com/yeslogic/allsorts/blob/332e29d8667f928bceb123833c139e885226399f/src/tables/cmap/subset.rs#L54) enum used to drive some of the `cmap` generation behaviour. A new variant could be added to that and then some way to select that strategy could be added.

**Comment by dnlmlr** [comment]
Yeah I agree that it shouldn't be default, since this is kind of an edge case. What I meant was a way to externally change the encoding mode, for example as a parameter to the `subset` function. Basically any mechanism that would allow to optionally prevent encoding with `MacRoman`.

**Comment by dnlmlr** [comment]
Another thing I noticed that would be extremely useful, is returning the actual GID mappings with the new subset font. That would make it very easy to map exiting GIDs to the new GIDs of the subset font and thereby updating "text" that is present as a stream of GIDs for the original font
